XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/MixedRealityFrameDiagnosticsDataProvider.cs

```diff
@@ -0,0 +1,177 @@
+            // Increase internal frame count since last FPS calculation.
+            ++frameCount;
+
+            // Do we have enough sample data for a new FPS calculation?
+            float elapsedSeconds = stopwatch.ElapsedMilliseconds * 0.001f;
+            if (elapsedSeconds >= frameSampleRate)
+            {
+                int newCPUFrameRate = (int)(1.0f / (elapsedSeconds / frameCount));
+                int newGPUFrameRate = 0;
+
+                // Many platforms do not yet support the FrameTimingManager. When timing data is returned from the FrameTimingManager we will use
+                // its timing data, else we will depend on the stopwatch.
+                uint frameTimingsCount = FrameTimingManager.GetLatestTimings((uint)Mathf.Min(frameCount, maxFrameTimings), frameTimings);
+                if (frameTimingsCount != 0)
+                {
+                    AverageFrameTiming(frameTimings, frameTimingsCount, out var cpuFrameTime, out var gpuFrameTime);
+                    newCPUFrameRate = (int)(1.0f / (cpuFrameTime / frameCount));
+                    newGPUFrameRate = (int)(1.0f / (gpuFrameTime / frameCount));
+                }
+
+                if (CPUFrameRate != newCPUFrameRate)
+                {
+                    for (int i = 0; i < Handlers.Count; i++)
+                    {
+                        Handlers[i].OnFrameRateChanged(CPUFrameRate, newCPUFrameRate, false);
+                    }
+
+                    CPUFrameRate = newCPUFrameRate;
+                }
+
+                if (newGPUFrameRate != 0 && GPUFrameRate != newGPUFrameRate)
+                {
+                    for (int i = 0; i < Handlers.Count; i++)
+                    {
+                        Handlers[i].OnFrameRateChanged(GPUFrameRate, newGPUFrameRate, true);
+                    }
+
+                    GPUFrameRate = newGPUFrameRate;
+                }
+
+                // Update missed frames.
+                for (int i = missedFramesRange - 1; i > 0; --i)
+                {
+                    missedFrames[i] = missedFrames[i - 1];
+                }
+
+                // Ideally we would query a device specific API (like the HolographicFramePresentationReport) to detect missed frames.
+                // But, many of these APIs are inaccessible in Unity. Currently missed frames are assumed when the average cpuFrameRate 
+                // is under the target frame rate.
+                missedFrames[0] = newCPUFrameRate < ((int)(DeviceTargetRefreshRate) - 1);
+                for (int i = 0; i < Handlers.Count; i++)
+                {
+                    Handlers[i].OnMissedFramesChanged(missedFrames);
+                }
+
+                // Reset timers.
+                frameCount = 0;
+                stopwatch.Reset();
+                stopwatch.Start();
+            }
+        }
+
+        /// <inheritdoc />
+        public override void Destroy()
+        {
+            base.Destroy();
+            stopwatch.Reset();
```

Comment:
Why reset the stopwatch if we're destroying this instance?

Reply:
Because we can? Good spot. Does not make sense.